### PR TITLE
WIP/ENH: various cleanups and cat_epochs functionality

### PR DESCRIPTION
**Pull request by dengemann**
Makes it easier to handle high sample frequencies with many runs.

cc @agramfort @cmoutard

### CODE REVIEW

File: mne/epochs.py [hunk 1]
```diff
@@ -2147,3 +2147,45 @@ def add_channels_epochs(epochs_list, name='Unknown', add_eeg_ref=True,
...
+    """
+    out = epochs_list[0].copy()
+    out._data = out.get_data()
+    out.preload = True
+    for ii, epochs in enumerate(epochs_list[1:]):
+        _compare_epochs_infos(epochs.info, epochs_list[0].info, ii)
+        if not np.all(epochs.times == epochs_list[0].times):
+            raise ValueError('Epochs must have same times')
+
+        out._data = np.concatenate((out._data, epochs.get_data()), axis=0)
```
Comment (Eric89GXL) [inline mne/epochs.py]: Don't you need to concat the events list, too?

File: mne/epochs.py [hunk 2]
```diff
@@ -2147,3 +2147,45 @@ def add_channels_epochs(epochs_list, name='Unknown', add_eeg_ref=True,
...
+        list of Epochs instances to concatenate (in order).
+
+    Returns
+    -------
+    epochs : instance of Epochs
+        The result of the concatenation (first Epochs instance passed in).
+    """
+    out = epochs_list[0].copy()
+    out._data = out.get_data()
+    out.preload = True
```
Comment (Eric89GXL) [inline mne/epochs.py]: This is a bit of a bummer, but is at least a good starting point.

File: mne/epochs.py (hunk 1 above)
Comment (dengemann) [inline mne/epochs.py]: @Eric89GXL yeah you're right.

File: mne/epochs.py [hunk 3]
```diff
@@ -2147,3 +2147,45 @@ def add_channels_epochs(epochs_list, name='Unknown', add_eeg_ref=True,
...
+    for ii, epochs in enumerate(epochs_list[1:]):
+        _compare_epochs_infos(epochs.info, epochs_list[0].info, ii)
+        if not np.all(epochs.times == epochs_list[0].times):
+            raise ValueError('Epochs must have same times')
+
+        out._data = np.concatenate((out._data, epochs.get_data()), axis=0)
+    return out
```
Comment (agramfort) [inline mne/epochs.py]: we should return an EpochsArray that has no .raw attribute. It would be cleaner.

Comment (Eric89GXL) [inline mne/epochs.py]: We should probably deprecate the `.raw` attribute of Epochs anyway for consistency if `EpochsArray` doesn't have it. Helps unify the set of properties `Epochs` classes have.

**Comment by dengemann** [comment]
@Eric89GXL @agramfort addressed discussion + fixed .events

File: mne/epochs.py (hunk 1 above)
Comment (agramfort) [inline mne/epochs.py]: I would construct the EpochsArray at the end and avoid the hack of accessing epochs._data which is meant to be private.

Comment (dengemann) [inline mne/epochs.py]: Why that? It's private and it's straight forward to use it inside a function.

Comment (agramfort) [inline mne/epochs.py]: it's not clean. What if we change the private stuff. It's a bad example.

Comment (dengemann) [inline mne/epochs.py]: pushed.

File: mne/epochs.py [hunk 4]
```diff
@@ -2147,3 +2147,54 @@ def add_channels_epochs(epochs_list, name='Unknown', add_eeg_ref=True,
...
+    """
+    out = epochs_list[0]
+    data = out.get_data()
+    events = out.events
+    for ii, epochs in enumerate(epochs_list[1:]):
+        _compare_epochs_infos(epochs.info, epochs_list[0].info, ii)
+        if not np.all(epochs.times == epochs_list[0].times):
+            raise ValueError('Epochs must have same times')
+
+        data = np.concatenate((data, epochs.get_data()), axis=0)
```
Comment (agramfort) [inline mne/epochs.py]: nitpick I would concatenate at the end to avoid malloc at each for pass

Comment (dengemann) [inline mne/epochs.py]: Ok, improved.

File: mne/epochs.py [hunk 5]
```diff
@@ -2147,3 +2147,55 @@ def add_channels_epochs(epochs_list, name='Unknown', add_eeg_ref=True,
...
+               zip(info2['projs'], info1['projs'])):
+        raise ValueError('SSP projectors in epochs files must be the same')
+
+
+def concatenate_epochs(epochs_list):
+    """Concatenate a list of epochs into one epochs object
+
+    Parameters
+    ----------
+    raws : list
```
Comment (agramfort) [inline mne/epochs.py]: raws? :)

Comment (dengemann) [inline mne/epochs.py]: oh my god!

**Comment by agramfort** [comment]
thanks @dengemann !

**Pull request merged**